compiler/src/dmd/dcast.d

```diff
@@ -3229,24 +3229,61 @@ Lagain:
 
             if (t1.ty == Tclass && t2.ty == Tclass)
             {
+                /* t1 cannot be converted to t2, and vice versa
+                 */
                 TypeClass tc1 = t1.isTypeClass();
                 TypeClass tc2 = t2.isTypeClass();
 
-                /* Pick 'tightest' type
-                 */
-                ClassDeclaration cd1 = tc1.sym.baseClass;
-                ClassDeclaration cd2 = tc2.sym.baseClass;
-                if (cd1 && cd2)
+                /* accumulate results in this */
+                static struct Root
                 {
-                    t1 = cd1.type.castMod(t1.mod);
-                    t2 = cd2.type.castMod(t2.mod);
+                    ClassDeclaration cd;
+                    int depth;
+                    bool error;
+
+                    /* merge cd into results */
+                    void merge(ClassDeclaration cd, int depth)
+                    {
+                        if (depth < this.depth) // less depth means better match
+                        {
+                            this.cd = cd;
+                            this.depth = depth;
+                            error = false;
+                        }
+                        else if (depth == this.depth)
+                        {
+                            error = this.cd != cd; // better be same if same depth
+                        }
+                        /* else deeper - ignore */
+                    }
                 }
-                else if (cd1)
-                    t1 = cd1.type;
-                else if (cd2)
-                    t2 = cd2.type;
-                else
-                    return null;
+
+                /* Find common root of cd1 and cd2. Accumulate results in r. depth is nesting level */
+                static void findCommonRoot(ClassDeclaration cd1, ClassDeclaration cd2, ref Root r, int depth)
+                {
+                    /* Warning: quadratic time function
```

Comment:
This function needs to be memoized, it currently has exponential time complexity. I wouldn't surprised if this causes the time-outs in the test suite.

Reply:
I'm pretty sure that is the problem. This one works, but is unacceptably slow. Anyone know of a better algorithm?

Reply:
Ok, I've thought of a better algorithm. It'll take some time to implement it.

Reply:
Does it have to be constant space? I don't see how you solve it efficiently without some kind of record.

I'd basically just do A* to get reachability for each of the roots of one symbol, then do reachability on the second root, finding any common ones via lookup.